From c495b9d3a3778a8500a494ec47d10dfb86c2522a Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:08:12 -0400
Subject: [powerpc] iommu: Don't detach device without IOMMU group

Message-id: <1410545655-205645-264-git-send-email-dzickus@redhat.com>
Patchwork-id: 94327
O-Subject: [RHEL7 PATCH 263/626] powerpc/iommu: Don't detach device without IOMMU group
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 0c4b9e27b09eeb4da84451c038a587b92ce93ff5
Author: Gavin Shan <shangw@linux.vnet.ibm.com>
Date:   Mon Jan 13 11:36:22 2014 +0800

    powerpc/iommu: Don't detach device without IOMMU group

    Some devices, for example PCI root port, don't have IOMMU table and
    group. We needn't detach them from their IOMMU group. Otherwise, it
    potentially incurs kernel crash because of referring NULL IOMMU group
    as following backtrace indicates:

      .iommu_group_remove_device+0x74/0x1b0
      .iommu_bus_notifier+0x94/0xb4
      .notifier_call_chain+0x78/0xe8
      .__blocking_notifier_call_chain+0x7c/0xbc
      .blocking_notifier_call_chain+0x38/0x48
      .device_del+0x50/0x234
      .pci_remove_bus_device+0x88/0x138
      .pci_stop_and_remove_bus_device+0x2c/0x40
      .pcibios_remove_pci_devices+0xcc/0xfc
      .pcibios_remove_pci_devices+0x3c/0xfc

    Signed-off-by: Gavin Shan <shangw@linux.vnet.ibm.com>
    Reviewed-by: Alexey Kardashevskiy <aik@ozlabs.ru>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/iommu.c b/arch/powerpc/kernel/iommu.c
index e35db96..b3b2e4c 100644
--- a/arch/powerpc/kernel/iommu.c
+++ b/arch/powerpc/kernel/iommu.c
@@ -1144,6 +1144,17 @@ EXPORT_SYMBOL_GPL(iommu_add_device);
 
 void iommu_del_device(struct device *dev)
 {
+ /*
+  * Some devices might not have IOMMU table and group
+  * and we needn't detach them from the associated
+  * IOMMU groups
+  */
+ if (!dev->iommu_group) {
+  pr_debug("iommu_tce: skipping device %s with no tbl\n",
+    dev_name(dev));
+  return;
+ }
+
  iommu_group_remove_device(dev);
 }
 EXPORT_SYMBOL_GPL(iommu_del_device);
-- 
1.7.1